asmcomp/asmgen.ml

```diff
@@ -12,12 +12,22 @@
 
 (* From lambda to assembly code *)
 
+[@@@ocaml.warning "+a-4-9-40-41-42"]
+
 open Format
 open Config
 open Clflags
 open Misc
 open Cmm
 
+type lambda_program =
+  { code : Lambda.lambda;
+    main_module_block_size : int; }
+
+type _ backend_kind =
```

Comment:
The use of a GADT here is overly complicated.  Simply expose two versions of `compile_implementation`.  Each call site seems knows if it is flambda or non-flambda.